runtime/onert/backend/trix/DevContext.cc

```diff
@@ -0,0 +1,265 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "DevContext.h"
+
+#include "Convert.h"
+
+namespace onert
+{
+namespace backend
+{
+namespace trix
+{
+
+DevContext::DevContext() : _dev_handles{}, _model_ids{}, _meta_map{}
+{
+  auto dev_count = getnumNPUdeviceByType(NPUCOND_TRIV2_CONN_SOCIP);
+  if (dev_count <= 0)
+  {
+    throw std::runtime_error("Unable to find TRIX NPU device");
+  }
+
+  // Get NPU device handles
+  for (int i = 0; i < dev_count; ++i)
+  {
+    npudev_h handle;
+    if (getNPUdeviceByType(&handle, NPUCOND_TRIV2_CONN_SOCIP, i) < 0)
+    {
+      throw std::runtime_error("Failed to get TRIX NPU device handle");
+    }
+    _dev_handles.emplace_back(handle);
+  }
+
+  // NOTE Do not change the number of threads as long as jobs in thread call
+  //      the synchronous APIs such as submitNPU_request()
+  _batch_thread_pool = std::make_unique<BatchThreadPool>(_dev_handles.size());
+}
+
+DevContext::~DevContext()
+{
+  // NOTE Must release _batch_thread_pool before releasing _dev_handles to wait for all threads to
+  //      be terminated
+  _batch_thread_pool.reset(nullptr);
+
+  for (const auto &dev_handle : _dev_handles)
+  {
+    unregisterNPUmodel_all(dev_handle);
+    putNPUdevice(dev_handle);
+  }
+}
+
+ModelID DevContext::registerModel(const std::string &model_file_path)
+{
+  auto meta = getNPUmodel_metadata(model_file_path.c_str(), false);
+
+  if (meta == nullptr)
+  {
+    throw std::runtime_error("Unable to extract the model metadata");
+  }
+
+  generic_buffer file_info;
+  file_info.type = BUFFER_FILE;
+  file_info.filepath = model_file_path.c_str();
+  file_info.size = meta->size;
+
+  ModelID model_id;
+
+  for (uint32_t dev_num = 0; dev_num < _dev_handles.size(); ++dev_num)
+  {
+    // Register model for each device
+    uint32_t model_id_at_device;
+    if (registerNPUmodel(_dev_handles.at(dev_num), &file_info, &model_id_at_device) < 0)
+    {
+      throw std::runtime_error("Failed to register npu model");
+    }
+
+    if (dev_num == 0)
+    {
+      model_id = model_id_at_device;
+      _meta_map[model_id_at_device] = std::shared_ptr<npubin_meta>(meta);
+    }
+    else
+    {
+      _meta_map[model_id_at_device] = _meta_map[model_id];
+    }
+
+    _model_ids[model_id].resize(dev_num + 1);
+    _model_ids[model_id].at(dev_num) = model_id_at_device;
+  }
+
+  // Return the model id for device 0 only
+  return model_id;
+}
+
+void DevContext::unRegisterModel(ModelID model_id)
+{
+  for (uint32_t dev_num = 0; dev_num < _dev_handles.size(); ++dev_num)
+  {
+    const auto model_id_at_device = _model_ids.at(model_id).at(dev_num);
+    const auto &dev_handle = _dev_handles.at(dev_num);
+
+    // Remove meta data
+    _meta_map.erase(model_id_at_device);
+
+    // Unregister Model for each device
+    unregisterNPUmodel(dev_handle, model_id_at_device);
```

Comment:
@ragmani Have you checked that this results in warning on x86 simulator ?

Reply:
The warning is printed on x86 simulator.